robustirc/src/jepsen/robustirc.clj

```diff
@@ -0,0 +1,230 @@
+  (reify db/DB
+    (setup! [this test node]
+      (c/su
+        (c/ssh* {:cmd "killall robustirc"})
+	(try (c/exec :dpkg-query :-l :golang-go)
+	     (catch RuntimeException _
+	       (info "Installing golang-go")
+	       (c/exec :apt-get :install :-y :golang-go)))
+	(try (c/exec :dpkg-query :-l :mercurial)
+	     (catch RuntimeException _
+	       (info "Installing mercurial")
+	       (c/exec :apt-get :install :-y :mercurial)))
+	(c/exec :env "GOPATH=~/gocode"
+	        :go :get :-u "github.com/robustirc/robustirc")
+
+        ; generated by resources/gencert.go
+        (c/upload (.getFile (io/resource "cert.pem")) "/tmp/cert.pem")
+        (c/upload (.getFile (io/resource "key.pem")) "/tmp/key.pem")
+        (c/ssh* {:cmd "rm -rf /var/lib/robustirc"})
+        (c/ssh* {:cmd "mkdir -p /var/lib/robustirc"})
+        (jepsen/synchronize test)
+
+        (let [cmd (str
+                "/sbin/start-stop-daemon --start --background --exec ~/gocode/bin/robustirc --"
+                " -listen=" (name node) ":13001"
+                " -network_password=secret"
+                " -network_name=jepsen"
+                " -tls_cert_path=/tmp/cert.pem"
+                " -tls_ca_file=/tmp/cert.pem"
+                " -tls_key_path=/tmp/key.pem"
+                " -singlenode")]
+          (if-not (= node (jepsen/primary test))
+                (Thread/sleep 1000)
+                (do
+                  (info node (str "running: " cmd))
+                  (c/ssh* {:cmd cmd})
+                  (Thread/sleep 5000))))
+        (jepsen/synchronize test)
+        (let [cmd (str
+                "/sbin/start-stop-daemon --start --background --exec ~/gocode/bin/robustirc --"
+                " -listen=" (name node) ":13001"
+                " -network_password=secret"
+                " -network_name=jepsen"
+                " -tls_cert_path=/tmp/cert.pem"
+                " -tls_ca_file=/tmp/cert.pem"
+                " -tls_key_path=/tmp/key.pem"
+                " -join=" (name (jepsen/primary test)) ":13001")]
+          (if (= node (jepsen/primary test))
+                (Thread/sleep 100)
+                (do
+                  (info node (str "running: " cmd))
+                  (c/ssh* {:cmd cmd})
+                  (Thread/sleep 5000))))
+        (jepsen/synchronize test)
+        (info node "setup done")))
+
+    (teardown! [this test node]
+      (c/su
+        (c/ssh* {:cmd "killall robustirc"}))
+      )))
+
+(defn with-nemesis
+  "Wraps a client generator in a nemesis that induces failures and eventually
+  stops."
+  [client]
+  (gen/phases
+    (gen/phases
+      (->> client
+           (gen/nemesis
+             (gen/seq (cycle [(gen/sleep 0)
+                              {:type :info, :f :start}
+                              (gen/sleep 10)
+                              {:type :info, :f :stop}])))
+           (gen/time-limit 30))
+      (gen/nemesis (gen/once {:type :info, :f :stop}))
+      (gen/sleep 5))))
+
+(defn createsession
```

Comment:
(I'm gonna merge, but just as a heads-up idiomatic Clojure style is to use hyphens, e.g. `create-session`. :))